### PR TITLE
Feature/profile ui

### PR DESCRIPTION
**Pull request by yazhyvyi**

### CODE REVIEW

File: client/src/components/Home.jsx [hunk 1]
```diff
@@ -1,13 +1,25 @@
...
-function Home() {
-  return (
-    <Row>
-      <Col s={4} offset="s2"><h1>This is Home</h1></Col>
-    </Row>
-  );
-}
+export default class Home extends Component {
+  componentDidMount() {
+    fetch(server_url)
```
Comment (Merteg) [inline client/src/components/Home.jsx]: This useless.

File: client/src/components/Nav/LogIn.jsx [hunk 2]
```diff
@@ -25,6 +25,7 @@ class LogIn extends Component {
     }
     request('/log-in', 'POST', JSON.stringify(log_in_data))
     .then(data=>{
+      console.log(data)
```
Comment (Merteg) [inline client/src/components/Nav/LogIn.jsx]: Please delete all console.log

File: client/src/components/Nav/NavProfile.jsx [hunk 3]
```diff
@@ -11,12 +11,20 @@ class NavProfile extends Component {
     else return '/person.jpg'
   }
 
+  constuctor() {
+    this.routeChange = this.routeChange.bind(this);
+   }
+
+   routeChange(){
```
Comment (Merteg) [inline client/src/components/Nav/NavProfile.jsx]: I know better way to do this, Sorry but seems really bad

File: client/src/components/Profile.jsx [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        };
+
+        this.UpdateClick = this.UpdateClick.bind(this);
+        this.onChangeHandler = this.onChangeHandler.bind(this);
+        this.handleClick_Main = this.handleClick_Main.bind(this);
+        this.handleClick_Edit = this.handleClick_Edit.bind(this);
+    }
+
+    componentDidMount() {
+        fetch(server_url + '/profile/' + window.location.href.split("/").pop())
```
Comment (Merteg) [inline client/src/components/Profile.jsx]: Why you don't use request function from utils.js?

File: client/src/components/Profile.jsx [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        .then(response => response.json())
+        .then(json => {
+            console.log(json);
+            this.setState({ user: json });
+        })
+    }
+
+    componentWillUnmount() {
+        clearInterval(this.timer);
```
Comment (Merteg) [inline client/src/components/Profile.jsx]: What is this?

File: client/src/components/Profile.jsx [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+          <h1>Profile</h1>
+        </div>
+        <ColoredLine color="teal" />
+        {page}
+      </div>
+    );
+  }
+}
+
+const ColoredLine = ({ color }) => (
```
Comment (Merteg) [inline client/src/components/Profile.jsx]: I don't think you need to define a new component. Also NEVER write the css properties here. Use .css file for this

File: client/src/components/Profile.jsx [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+    handleClick_Main() {
+        window.location.reload();
+    }
+    handleClick_Edit() {
+        this.setState({Edit: true});
+    }
+
+  render() {
+      let avatar_image;
+      if (this.state.user.avatar == null) {
```
Comment (Merteg) [inline client/src/components/Profile.jsx]: Please remove all statements from render, better to create new method (smth like avatarImage(){})

File: client/src/components/Profile.jsx [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+    }
+
+    onChangeHandler = (e) => {
+        let { id } = e.currentTarget
+        this.setState({ [id]: e.currentTarget.value })
+    }
+
+    handleClick_Main() {
+        window.location.reload();
```
Comment (Merteg) [inline client/src/components/Profile.jsx]: when you change the `state`, the component instantly re-render. You don't need to reload whole page.

File: client/src/components/Profile.jsx [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+import { Row, Col, Input, Button } from 'react-materialize';
+import { server_url } from '../config.json';
+
+
+export default class Profile extends Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            Edit: false,
```
Comment (Merteg) [inline client/src/components/Profile.jsx]: why uppercase?

File: client/src/components/Profile.jsx [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+            msg: '',
+            user: [],
+        };
+
+        this.UpdateClick = this.UpdateClick.bind(this);
```
Comment (Merteg) [inline client/src/components/Profile.jsx]: You don't pass this methods to another components so you don't need to bind this (and call constructor too)

File: client/src/index.js [hunk 11]
```diff
@@ -1,6 +1,6 @@
 import React from 'react';
 import ReactDOM from 'react-dom';
-import App from './App';
+import App from './App.jsx';
```
Comment (Merteg) [inline client/src/index.js]: You don't need to write extension of the component

File: requirements.txt [hunk 12]
```diff
@@ -4,6 +4,7 @@ freeze==1.0.10
 hupper==1.3
 Mako==1.0.7
 MarkupSafe==1.0
+npm==0.1.1
```
Comment (Merteg) [inline requirements.txt]: WHAT?

File: client/src/components/Home.jsx [hunk 13]
```diff
@@ -1,13 +1,25 @@
...
+      .then(data => {
+        console.log('Data: ', data)
+      })
+  }
 
-export default Home;
+  render() {
+    return (
+      <Row>
+        <Col s={1}></Col>
```
Comment (RayMet) [inline client/src/components/Home.jsx]: Why first "Col" is empty?

File: server/development.ini.sample [hunk 14]
```diff
@@ -1,68 +0,0 @@
-###
```
Comment (Merteg) [inline server/development.ini.sample]: We all need this file, please don't delete it

**Review by RayMet — CHANGES_REQUESTED** [review]

File: server/server/models/user.py [hunk 15]
```diff
@@ -81,9 +81,10 @@ def get_role(self):
         return self.roles.role
 
     @staticmethod
-    def update_user(json_data, request):
+    def update_user(request, json_data):
         """ Method to update user in database """
-        if request.dbsession.query(User).get(request.matchdict['profile_id']):
```
Comment (Merteg) [inline server/server/models/user.py]: this line is not longer than 79 symbols

File: server/server/views/home.py [hunk 16]
```diff
@@ -0,0 +1,20 @@
+from cornice.resource import resource, view
+from pyramid.security import Allow, Everyone
+
+
+@resource(path='/', renderer='json', cors_origins=('*',))
```
Comment (Merteg) [inline server/server/views/home.py]: Please don't allow cors to all, you need to allow it just for client

File: server/server/views/home.py [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+class Home(object):
+
+    def __init__(self, request, context=None):
+        self.request = request
+        self.context = context
+
+    def __acl__(self):
+        return [(Allow, Everyone, 'everything')]
```
Comment (Merteg) [inline server/server/views/home.py]: 'everything' sounds really bad, pyramid have ALL_PERMISSION const for this

File: server/server/views/profile.py [hunk 18]
```diff
@@ -1,3 +1,12 @@
+import json
+import transaction
+
+from cornice.resource import resource, view
+from cornice import Service
```
Comment (Merteg) [inline server/server/views/profile.py]: Remove all unused imports

File: server/server/views/profile.py [hunk 19]
```diff
@@ -6,23 +15,29 @@
...
 
-    def __init__(self, request, context=None):
+    def __init__(self, request, context=None,):
         self.request = request
         self.context = context
 
     def __acl__(self):
         return [(Allow, Everyone, ALL_PERMISSIONS),
-                (Allow, self.owner, 'edit')]
+                (Allow, Authenticated, 'edit')]
```
Comment (Merteg) [inline server/server/views/profile.py]: It means that the every authenticated user can edit every profile page. It does not have to work like that

**Review by Merteg — CHANGES_REQUESTED** [review]

File: client/src/components/Home.jsx (hunk 13 above)
Comment (yazhyvyi) [reply client/src/components/Home.jsx]: ok, i'll correct that

File: client/src/components/Profile.jsx (hunk 4 above)
Comment (yazhyvyi) [reply client/src/components/Profile.jsx]: I wrote this piece of code before utils.js was implemented

File: client/src/components/Profile.jsx (hunk 5 above)
Comment (yazhyvyi) [reply client/src/components/Profile.jsx]: Well, forgot to remove that

**Comment by yazhyvyi** [comment]
Update

**Review by ykobuk — APPROVED** [review]

File: client/src/components/Nav/NavProfile.jsx (hunk 3 above)
Comment (yazhyvyi) [reply client/src/components/Nav/NavProfile.jsx]: Care to elaborate?

File: server/server/views/profile.py [hunk 20]
```diff
@@ -1,28 +1,38 @@
...
-        return model_to_dict(user)
+        user_dict = model_to_dict(user)
+        user_dict['password'] = None
+        return user_dict
 
+    @view(permission='edit')
     def put(self):
         request = self.request
         data = request.json_body
+        if 'password' in data is not "":
```
Comment (serge30) [inline server/server/views/profile.py]: First, I was going to write that this statement returns incorrect result, but after experiments in console it works. What is weird.

Anyway correct statement to check key presence in the dict is `if 'password' in data`.

Comment (serge30) [reply server/server/views/profile.py]: or if you need to check whether key is present in dict and value is not empty, you may try: `if dict.get('password'):`

Comment (yazhyvyi) [reply server/server/views/profile.py]: Nope, returns error: 
TypeError: descriptor 'get' requires a 'dict' object but received a 'str'

Edit: Nevermind, it works, mistake on my side

Comment (serge30) [reply server/server/views/profile.py]: I provide a wrong example, you need to use your variable:
```
if data.get('password'):
```

File: client/src/components/Nav/NavProfile.jsx (hunk 3 above)
Comment (Merteg) [reply client/src/components/Nav/NavProfile.jsx]: `<li><Link to={"/profile/" + sessionStorage['User-id']}>` seems better and also takes less code and resources

File: client/src/components/Profile.jsx (hunk 4 above)
Comment (Merteg) [reply client/src/components/Profile.jsx]: But now it's here, it takes less space and seems better. Also if we will need to upgrade it, better to do this in one place.

File: client/src/components/Profile.jsx [hunk 21]
```diff
@@ -0,0 +1,144 @@
...
+        this.handleClick_Main = this.handleClick_Main.bind(this);
+        this.handleClick_Edit = this.handleClick_Edit.bind(this);
+    }
+
+    componentDidMount() {
+        fetch(server_url + '/profile/' + window.location.href.split("/").pop())
+        .then(response => response.json())
+        .then(json => {
+            this.setState({ user: json });
+            console.log(json)
```
Comment (Merteg) [inline client/src/components/Profile.jsx]: Remove console.log...

File: client/src/components/Profile.jsx (hunk 4 above)
Comment (Merteg) [reply client/src/components/Profile.jsx]: Also I don't understand `window.location.href.split("/").pop()`, you define this route like ` <Route path="/profile/:profile_id" component={Profile}/>` so the path parameter stored in this.props.match.params.profile_id

File: client/src/components/Routes.jsx [hunk 22]
```diff
@@ -1,11 +1,13 @@
 import React from 'react';
 import { Route } from 'react-router-dom';
-import Home from './Home';
+import Home from './Home.jsx';
```
Comment (Merteg) [inline client/src/components/Routes.jsx]: Why do you type the extension of files???

File: server/server/views/home.py [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+class Home(object):
+
+    def __init__(self, request, context=None):
+        self.request = request
+        self.context = context
+
+    def __acl__(self):
+        return [(Allow, Everyone, ALL_PERMISSIONS)]
+
+    @view(permission=ALL_PERMISSIONS)
```
Comment (Merteg) [inline server/server/views/home.py]: Useless. We don't need any requests on the home page and this code also do nothing useful.

File: server/server/views/profile.py [hunk 24]
```diff
@@ -1,28 +1,38 @@
-from cornice.resource import resource
+import json
```
Comment (Merteg) [inline server/server/views/profile.py]: Do you use this import?

**Review by Merteg — CHANGES_REQUESTED** [review]

File: client/src/components/Profile.jsx (hunk 21 above)
Comment (YaZhyvyi) [reply client/src/components/Profile.jsx]: About NavProfile's routeChange:

I just tested your solution while being In the profile itself and it returns this error:
Warning: Hash history cannot PUSH the same path; a new entry will not be added to the history stack

File: client/src/components/Nav/NavProfile.jsx [hunk 25]
```diff
@@ -1,6 +1,7 @@
 import React, { Component } from 'react';
 import { NavItem } from 'react-materialize'
 import {isLogged} from '../../utils'
+import {Link} from 'react-router-dom'
```
Comment (Merteg) [inline client/src/components/Nav/NavProfile.jsx]: In accordance to style code [Add spaces inside curly braces](https://github.com/airbnb/javascript/#whitespace--in-braces)

File: server/server/views/home.py [hunk 26]
```diff
@@ -1,5 +1,4 @@
 from cornice.resource import resource, view
```
Comment (Merteg) [inline server/server/views/home.py]: We don't have this file in views and we don't need it right now. Please remove this.

File: server/server/views/profile.py [hunk 27]
```diff
@@ -1,28 +1,36 @@
...
         self.request = request
         self.context = context
+        self.owner_id = int(request.matchdict['profile_id'])
 
     def __acl__(self):
         return [(Allow, Everyone, ALL_PERMISSIONS),
-                (Allow, self.owner, 'edit')]
+                (Allow, self.owner_id, 'edit')]
 
+    @view(permission=ALL_PERMISSIONS)
```
Comment (Merteg) [inline server/server/views/profile.py]: It not a problem but if you allow it for everyone, you don't need to add permission at all.
I had to say this in the previous comment, sorry.

File: server/server/views/home.py [hunk 28]
```diff
@@ -0,0 +1,15 @@
+from cornice.resource import resource, view
```
Comment (Merteg) [inline server/server/views/home.py]: Maybe I typed this comment earlier, nevermind...
We don't need this file at all, we don't make any requests to this right now

File: client/src/components/Profile.jsx [hunk 29]
```diff
@@ -0,0 +1,143 @@
...
+        );
+      }
+      else {
+          return (
+            <div>
+              <div>
+                <h1>Profile</h1>
+              </div>
+                <Row>
+                    <Input s={12} id="email" label="Email" placeholder={this.state.user.email} onChange={this.onChangeHandler} />
```
Comment (Merteg) [inline client/src/components/Profile.jsx]: The input at Navbar have id="email" too. Id must be unique, you can name it smth like email_profile...

File: client/src/components/Profile.jsx [hunk 30]
```diff
@@ -0,0 +1,143 @@
...
+          return (
+            <div>
+              <div>
+                <h1>Profile</h1>
+              </div>
+                <div id="main">
+                <Row>
+                    <Col s={6}>
+                        <Row>
+                            <this.Avatar_image />
```
Comment (Merteg) [inline client/src/components/Profile.jsx]: Seems strange, you can call it like a simple method {this.avatar_image}
Also better to name method smth like `renderAvatarImage`.
Better to use camelCase in JS [style code](https://github.com/airbnb/javascript/#naming--camelCase)

**Review by Merteg — CHANGES_REQUESTED** [review]
It seems good. I type a comments about minor but I think important changes.

**Comment by yazhyvyi** [comment]
Update

File: server/server/views/home.py [hunk 31]
```diff
@@ -1,7 +1,3 @@
-from cornice.resource import resource, view
-
-
-@resource(path='/', renderer='json', cors_origins=('http://localhost:3000',))
 class Home(object):
```
Comment (Merteg) [inline server/server/views/home.py]: Remove. This. File.
Just DELETE.

File: client/src/components/Profile.jsx [hunk 32]
```diff
@@ -12,7 +12,7 @@ export default class Profile extends Component {
             user: []
         };
 
-        this.Avatar_image = this.Avatar_image.bind(this);
+        this.renderAvatarImage = this.renderAvatarImage.bind(this);
         this.UpdateClick = this.UpdateClick.bind(this);
         this.onChangeHandler = this.onChangeHandler.bind(this);
         this.handleClick_Main = this.handleClick_Main.bind(this);
```
Comment (Merteg) [inline client/src/components/Profile.jsx]: This don't have any sense here

**Review by Merteg — CHANGES_REQUESTED** [review]

**Comment by yazhyvyi** [comment]
Update

File: client/src/components/Profile.jsx [hunk 33]
```diff
@@ -83,7 +83,7 @@ export default class Profile extends Component {
                 <Row>
                     <Col s={6}>
                         <Row>
-                            <renderAvatarImage />
+                            <this.renderAvatarImage />
```
Comment (Merteg) [inline client/src/components/Profile.jsx]: {this.renderAvatarImage()}

**Review by Merteg — CHANGES_REQUESTED** [review]
Also you don't change fetch to request

**Comment by yazhyvyi** [comment]
Update

**Comment by Merteg** [comment]
Fetch - request??? If you think that write `fetch` is better tell me why?

**Comment by yazhyvyi** [comment]
Update

**Review by Merteg — APPROVED** [review]

**Pull request merged**